### PR TITLE
udev: Add support to configure Generic Segment Offload

### PR DESCRIPTION
**Pull request by ssahani**

### CODE REVIEW

File: src/network/networkd-network.h [hunk 1]
```diff
@@ -92,6 +92,8 @@ struct Network {
         struct ether_addr *mac;
         uint32_t mtu;
         uint32_t group;
+        uint32_t gso_max_size;
```
Comment (yuwata) [inline src/network/networkd-network.h]: `config_parse_iec_size()` takes `size_t`. Please use `size_t` for `gso_max_size`.

**Review by yuwata — COMMENTED** [review]
I am not sure but maybe these settings should be moved to .link file, instead of .network file.

File: src/network/networkd-network-gperf.gperf [hunk 2]
```diff
@@ -59,6 +59,8 @@ Match.Architecture,                          config_parse_net_condition,
 Link.MACAddress,                             config_parse_hwaddr,                                      0,                             offsetof(Network, mac)
 Link.MTUBytes,                               config_parse_mtu,                                         AF_UNSPEC,                     offsetof(Network, mtu)
 Link.Group,                                  config_parse_uint32,                                      0,                             offsetof(Network, group)
+Link.GenericSegmenOffloadMaxBytes,           config_parse_iec_size,                                    0,                             offsetof(Network, gso_max_size)
```
Comment (yuwata) [inline src/network/networkd-network-gperf.gperf]: Segmen -> Segment

**Comment by ssahani** [comment]
> I am not sure but maybe these settings should be moved to .link file, instead of .network file.

These are same as group/multicast/allmulticast etc. Hence it should be here. Thanks for the review .ptal

**Comment by yuwata** [comment]
> > I am not sure but maybe these settings should be moved to .link file, instead of .network file.
> 
> These are same as group/multicast/allmulticast etc. Hence it should be here. Thanks for the review .ptal

In what sense "these are the same"? .link file already has `GenericSegmentationOffload=` boolean setting. Does the setting have no correspondence to the new settings?

**Comment by ssahani** [comment]
> In what sense "these are the same"? .link file already has `GenericSegmentationOffload=` boolean setting. Does the setting have no correspondence to the new settings?

`.link` file turns on the tx-generic-segmentation feature. Here  are setting 

gso_max_size:  Maximum size of generic segmentation offload
gso_max_segs:  Maximum number of segments that can be passed to the NIC for GSO

What is your confusion ?

**Comment by yuwata** [comment]
> > In what sense "these are the same"? .link file already has `GenericSegmentationOffload=` boolean setting. Does the setting have no correspondence to the new settings?
> 
> `.link` file turns on the tx-generic-segmentation feature. Here are setting
> 
> gso_max_size: Maximum size of generic segmentation offload
> gso_max_segs: Maximum number of segments that can be passed to the NIC for GSO
> 
> What is your confusion ?

I do not have enough knowledge about them. So, please let me know. Are these options corresponding to each other? E.g., what happens if `tx-generic-segmentation` is disabled? Are `gso_max_size` and `gso_max_segs` ignored in that case? Or these `gso_max` settings are independent of `tx-generic-segmentation`?

**Comment by ssahani** [comment]
>I do not have enough knowledge about them. So, please let me know. Are these options corresponding to each other? E.g., what happens if tx-generic-segmentation is disabled? Are gso_max_size and gso_max_segs ignored in that case? Or these gso_max settings are independent of tx-generic-segmentation?

```GenericSegmentationOffload=``` I should have written offload . It means  physical device’s TX offloading capabilities.  Device driver can offload or not. Not like `tx-generic-segmentation` off.  As per I know gso_max_size and gso_max_segs will *not* be ignored.

**Comment by ssahani** [comment]
Cc @bluca

**Comment by bluca** [comment]
Needs rebase.

Regarding where to place the setting, I think the current approach makes sense: tx-generic-segmentation/GenericSegmentationOffload are like other settings managed via ethtool in the old ways - they are device driver settings, so they get configured by udev.
gso_max_size/segs on the other hand are interface settings, manged via iproute2, so it makes sense that they handled by networkd.

At least this is my understanding from looking at ethtool/iproute2 interfaces and documentation about the features. I have not looked at the kernel code.

**Comment by ssahani** [comment]
rebased thanks fir the review.

**Comment by yuwata** [comment]
Still I am not sure this should be added to .network or .link. But, at least, please rebase this PR.

**Comment by ssahani** [comment]
rebased ptal thanks !

**Comment by poettering** [comment]
This really looks like something that should be in .link, not .network

I mean the effect of this is not really visible to the network is it? It's an implementation detail of the server that should be invisble outside? If that's the case then it should be a .link option, not a .network one.

**Comment by poettering** [comment]
(i mean, the line between what belongs in .network and what in .link is blurry, and there are options belonging in both ultimately, because they are a property of the local hw that has effects on the network, but as I understand GSO it's primarily an implementation detail locally with no visibility outside of the local machine hence in this case it appears pretty clear to me this falls onto the .link side to me)

**Comment by ssahani** [comment]
updated thanks for the review. ptal

File: src/udev/net/link-config-gperf.gperf [hunk 3]
```diff
@@ -20,49 +20,51 @@ struct ConfigPerfItem;
...
+Link.RxMiniBufferSize,                 config_parse_nic_buffer_size,          0,                             offsetof(link_config, ring)
+Link.RxJumboBufferSize,                config_parse_nic_buffer_size,          0,                             offsetof(link_config, ring)
+Link.TxBufferSize,                     config_parse_nic_buffer_size,          0,                             offsetof(link_config, ring)
+Link.RxFlowControl,                    config_parse_tristate,                 0,                             offsetof(link_config, rx_flow_control)
+Link.TxFlowControl,                    config_parse_tristate,                 0,                             offsetof(link_config, tx_flow_control)
+Link.AutoNegotiationFlowControl,       config_parse_tristate,                 0,                             offsetof(link_config, autoneg_flow_control)
+Link.GenericSegmentOffloadMaxBytes,    config_parse_iec_size,                 0,                             offsetof(link_config, gso_max_size)
+Link.GenericSegmentOffloadMaxSegments, config_parse_uint32,                   0,                             offsetof(link_config, gso_max_segments)
```
Comment (yuwata) [inline src/udev/net/link-config-gperf.gperf]: Please add a line break at the end of the file.

File: test/fuzz/fuzz-link-parser/directives.link [hunk 4]
```diff
@@ -46,3 +46,5 @@ TxBufferSize=
 RxFlowControl=
 TxFlowControl=
 AutoNegotiationFlowControl=
+GenericSegmentOffloadMaxBytes=
+GenericSegmentOffloadMaxSegments=
```
Comment (yuwata) [inline test/fuzz/fuzz-link-parser/directives.link:50]: Please merge the last commit with the previous.

**Review by yuwata — CHANGES_REQUESTED** [review]
Mostly ok. Several nitpicks.

**Review by yuwata — APPROVED** [review]

**Pull request merged**